Feat/bpdm2 698 gate changelog endpoint

fabiodmota opened the pull request:
Add two new endpoints for Changelog with objects and logic on response 
Merge with main update

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/controller/ChangelogController.kt:
```diff
@@ -0,0 +1,87 @@
+@RestController
+@RequestMapping("/api/catena/business-partners/changelog")
+@Validated
+class ChangelogController (
```
The Controller should inherit from a ChangelogAPI that is defined in the BPDM Gate API module

Move all Springdoc annotations to the API interface

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/controller/ChangelogController.kt:
```diff
@@ -0,0 +1,87 @@
+        @ParameterObject @Valid paginationRequest: PaginationRequest,
+        @Parameter(description = "From Time", example = "2023-03-20T10:23:28.194Z") @RequestParam(required = false) fromTime: Instant?,
+        @RequestBody(required = true)  @NotEmpty(message = "Input externalIds list cannot be empty.") externalIds: Collection<String>
+    ): PageChangeLogResponse<ChangelogEntity> {
```
Dont respond with the whole entity here, use a DTO that will be part of the API model

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/controller/ChangelogController.kt:
```diff
@@ -0,0 +1,87 @@
+        @ParameterObject @Valid paginationRequest: PaginationRequest,
+        @Parameter(description = "From Time", example = "2023-03-20T10:23:28.194Z") @RequestParam fromTime: Instant?,
+        @Parameter(description = "LSA Type") @RequestParam lsaType: LsaType?
+    ): PageResponse<ChangelogEntity> {
```
Same as above, replace entity with a DTO fit for returning

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/controller/ChangelogController.kt:
```diff
@@ -0,0 +1,87 @@
+    fun getChangelogEntriesLsaType(
+        @ParameterObject @Valid paginationRequest: PaginationRequest,
+        @Parameter(description = "From Time", example = "2023-03-20T10:23:28.194Z") @RequestParam fromTime: Instant?,
+        @Parameter(description = "LSA Type") @RequestParam lsaType: LsaType?
```
LsaType is not a good enumeration to take here as it contains "None". Would need a new enumeration just containing the actual three LSA types

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/entity/ChangelogEntity.kt:
```diff
@@ -0,0 +1,40 @@
+    val externalId: String,
+    @Enumerated(EnumType.STRING)
+    @Column(name = "business_partner_type", nullable = false, updatable = false)
+    val businessPartnerType: LsaType
```
Same here with the LsaType. Better create a new enumeration which only contains the actual three LSA types. 
Suggestion: rename this LsaType to OptionalLsaType and introduce a new LsaType enumeration with only those three types as options.

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/exception/GateErrorCodes.kt:
```diff
@@ -33,3 +33,8 @@ enum class BusinessPartnerOutputError : ErrorCode {
+@Schema(description = "ChangeLogOutputError")
+enum class ChangeLogOutputError : ErrorCode {
+    ExternalIdNotFound,
+}
```
Needs to be moved to the BPDM Gate API module

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/repository/ChangelogRepository.kt:
```diff
@@ -0,0 +1,64 @@
+
+
+interface ChangelogRepository : JpaRepository<ChangelogEntity, Long> {
+
```
Repository methods here could be a good option for using the criteria library

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/repository/ChangelogRepository.kt:
```diff
@@ -0,0 +1,64 @@
+    //Query for ExternalID and fromTime Parameter
+    fun findAllByExternalIdInAndCreatedAtGreaterThanEqual(
+        externalIds: Collection<String>,
+        createdAt: Instant?,
```
Why nullable? Idea is that the createdAt timestamp needs to be there

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/repository/ChangelogRepository.kt:
```diff
@@ -0,0 +1,64 @@
+
+    //Query for BusinessPartnerType only
+    fun findAllByBusinessPartnerType(
+        businessPartnerType: LsaType?,
```
Same as above

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/repository/ChangelogRepository.kt:
```diff
@@ -0,0 +1,64 @@
+    fun findAllByBusinessPartnerTypeAndCreatedAtGreaterThanEqual(
+        businessPartnerType: LsaType?,
+        createdAt: Instant?,
+        pageable: Pageable
```
Same as above

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/ChangelogService.kt:
```diff
@@ -0,0 +1,97 @@
+@Service
+class ChangelogService(private val changelogRepository: ChangelogRepository) {
+
+
```
Logic here can be simplified with introduction of custom specifications.

Have a look at the PartnerChaneglogService in the BPDM Pool, specifically method: getChangelogEntriesByBpn

nicoprow commented on bpdm-gate/src/main/resources/application.properties:
```diff
@@ -20,7 +20,7 @@ bpdm.name=@project.name@
 bpdm.version=@project.version@
 #Change default port since 8080 is reserved for pool
-server.port=8081
+server.port=8082
```
Why change the port here?

nicoprow reviewed (changes requested):
Commits contain merge commits which should be removed from the history by rebasing.

Missing feature: add changelog entries when business partners are created/updated via PUT endpoint.

Missing tests

Missing Helm Chart adaptions (will be provided by me later)

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/LegalEntityService.kt:
```diff
@@ -40,11 +43,17 @@ class LegalEntityService(
+
+        legalEntities.forEach { legalEntity ->
+            changelogRepository.save(ChangelogEntity(legalEntity.externalId, LsaType.LegalEntity))
+        }
```
Better move the changelog after the actual upsert since the call to CDQ could go wrong

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/AddressService.kt:
```diff
@@ -148,6 +152,11 @@ class AddressService(
+
+        addresses.forEach { address ->
+            changelogRepository.save(ChangelogEntity(address.externalId, LsaType.Address))
+        }
```
Better move the changelog after the actual upsert since the call to CDQ could go wrong

nicoprow commented on bpdm-gate-api/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/api/model/response/ChangelogResponse.kt:
```diff
@@ -17,19 +18,17 @@
+    @Schema(description = "The type of the change")
+    val businessPartnerType: LsaType,
+    @Schema(description = "The timestamp of the operation")
+    val createdAt: Instant
```
Field name createdAt is a bit misleading here. The changelog entry should state when a business partner has been updated. In this regard better name the field soemthing like "modifiedAt", "updatedAt" or similar.

nicoprow commented on bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/ChangelogService.kt:
```diff
@@ -0,0 +1,89 @@
+            ErrorInfo(
+                ChangeLogOutputError.ExternalIdNotFound,
+                "$it not found",
+                "externalId not found"
```
Last argument here should only contain the actual external-id of the record in question

nicoprow reviewed (changes requested):
Works already well, just a few tweaks needed

nicoprow reviewed (changes requested):
Ok, content is good and ready to merge.

I just had a close look at the commit Message and I think there should be a bit of renaming here. Some of these messages are too generic if the context is missing. Either add a general context description as the main message and add the information in the detailed message under it or provide context with each commit message. The commit messages in question:

- [fix(gate): added externalid field](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/05e2af064414768490bc4fe90e04aeb7373d4f75)
- [fix(Gate): Fix Migration to map with entity](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/8e735b6b71fb2b343ef26c64b393401397ebdcda)
- [fix(gate): added externalid field](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/ffc166efa3728ceb983f2f351e76f0734d61ad19) is even a duplicate message
- [fix(Gate): Fix Migration to map with entity](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/86d7f736a835ddfef686b9db93c145b4895ef13a) also duplicate
- [Fix(Gate): Add missing logic](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/3bb42cecc0f73fff23ba5a1042d214f54a3afa90)
- [Fix(Gate): Fix Imports After rebase](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/2a9da7bf342f2fe184765aaef56bc12ab413a4e7)
- [Fix(Gate): Fix not accepting empty externalIds list](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/dac92f28d1ff67fe0bd2bc35acda87bf8a748945)
- [Fix(Gate): Fix port and use specifications on querys](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/8a12523f1b90c886580687ded41a7d943b043b11)
- [test(Gate): Add unit test for new controller](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/18a386d1061b5d3d39d756da7d42f8353897fc03)
- [fix(Gate): Fix imports](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/21c584ffcc8858d4f6bebc749196bf1e2974914a)
- [fix(Gate): Fix rebase](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/2280ce9949fd955356e02975c431ddcba2131177)
- [fix(Gate): Resolve and optimize code](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/1e7e19f6d955c50f39a4b8cf4bdfaf7efc9ced04)

Some of these commits are also not fixes but rather refactorings, so would better be better expressed with refactor(...): ...

Optional: Some of these commits may be squashed together, especially when there are several refactoring commits after each other. Can make a broader message and add details in the detail message

fabiodmota commented:
> Ok, content is good and ready to merge.
> 
> I just had a close look at the commit Message and I think there should be a bit of renaming here. Some of these messages are too generic if the context is missing. Either add a general context description as the main message and add the information in the detailed message under it or provide context with each commit message. The commit messages in question:
> 
> * [fix(gate): added externalid field](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/05e2af064414768490bc4fe90e04aeb7373d4f75)
> * [fix(Gate): Fix Migration to map with entity](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/8e735b6b71fb2b343ef26c64b393401397ebdcda)
> * [fix(gate): added externalid field](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/ffc166efa3728ceb983f2f351e76f0734d61ad19) is even a duplicate message
> * [fix(Gate): Fix Migration to map with entity](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/86d7f736a835ddfef686b9db93c145b4895ef13a) also duplicate
> * [Fix(Gate): Add missing logic](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/3bb42cecc0f73fff23ba5a1042d214f54a3afa90)
> * [Fix(Gate): Fix Imports After rebase](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/2a9da7bf342f2fe184765aaef56bc12ab413a4e7)
> * [Fix(Gate): Fix not accepting empty externalIds list](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/dac92f28d1ff67fe0bd2bc35acda87bf8a748945)
> * [Fix(Gate): Fix port and use specifications on querys](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/8a12523f1b90c886580687ded41a7d943b043b11)
> * [test(Gate): Add unit test for new controller](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/18a386d1061b5d3d39d756da7d42f8353897fc03)
> * [fix(Gate): Fix imports](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/21c584ffcc8858d4f6bebc749196bf1e2974914a)
> * [fix(Gate): Fix rebase](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/2280ce9949fd955356e02975c431ddcba2131177)
> * [fix(Gate): Resolve and optimize code](https://github.com/eclipse-tractusx/bpdm/pull/91/commits/1e7e19f6d955c50f39a4b8cf4bdfaf7efc9ced04)
> 
> Some of these commits are also not fixes but rather refactorings, so would better be better expressed with refactor(...): ...
> 
> Optional: Some of these commits may be squashed together, especially when there are several refactoring commits after each other. Can make a broader message and add details in the detail message

Done refactoring the commits

nicoprow reviewed (changes requested):
Ok, did some final testing and found some strange behaviour:

When you request changelogs via external-id:
When there is entry with the requested external-id on the page the Gate returns an error that the external-id is not found.However, the external-id exists, just that it's not on the current page. This behaviour need to be fixed
